### PR TITLE
Backoff failing RSS feeds

### PR DESCRIPTION
**Pull request by Half-Shot**
This should generally help speed up working feeds, by backing off failing ones. This is useful for environments with 1000s of feeds, some of which may have rotted since they were set up. For instance, integrations.ems.host sees about a 5% failure rate at the HTTP level alone.

Leftovers from this PR:
 - Cache backoff times across restarts in Redis. We currently hit all feeds on startup and backoff again. We would need to provide a flush button though.
 - General CPU improvements, now that we're failing less, the CPU is working harder.

### CODE REVIEW

File: src/util/mod.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        for item in self.backoff.iter() {
+            if *item.1 < since_the_epoch.as_millis() {
+                self.queue.push_back(item.0.clone());
+                items_to_rm.push(item.0.clone());
+            }
+        }
+
+        for item in items_to_rm {
+            self.backoff.remove(&item);
+        }
```
Comment (Half-Shot) [inline src/util/mod.rs:50]: Something about this strikes me as inefficient. I sort of want to effectively pop only some items from the HashMap. Currently we have to do a O(N) iteration on each pop, which is pretty rubbish.

File: src/Stores/StorageProvider.ts [hunk 2]
```diff
@@ -25,7 +25,7 @@ export interface IBridgeStorageProvider extends IAppserviceStorageProvider, ISto
     setStoredTempFile(key: string, value: string): Promise<void>;
     getGitlabDiscussionThreads(connectionId: string): Promise<SerializedGitlabDiscussionThreads>;
     setGitlabDiscussionThreads(connectionId: string, value: SerializedGitlabDiscussionThreads): Promise<void>;
-    storeFeedGuids(url: string, ...guid: string[]): Promise<void>;
-    hasSeenFeed(url: string, ...guid: string[]): Promise<boolean>;
-    hasSeenFeedGuid(url: string, guid: string): Promise<boolean>;
+    storeFeedGuids(url: string, ...guids: string[]): Promise<void>;
+    hasSeenFeed(url: string): Promise<boolean>;
+    hasSeenFeedGuids(url: string, ...guids: string[]): Promise<string[]>;
```
Comment (Half-Shot) [inline src/Stores/StorageProvider.ts:30]: Sped this up by checking multiple guids.

File: src/Stores/RedisStorageProvider.ts [hunk 3]
```diff
@@ -216,17 +215,26 @@ export class RedisStorageProvider extends RedisStorageContextualProvider impleme
...
     }
 
-    public async hasSeenFeedGuid(url: string, guid: string): Promise<boolean> {
-        return (await this.redis.lpos(`${FEED_GUIDS}${url}`, guid)) != null;
+    public async hasSeenFeedGuids(url: string, ...guids: string[]): Promise<string[]> {
+        let multi = this.redis.multi();
+        for (const guid of guids) {
+            multi = multi.lpos(`${FEED_GUIDS}${url}`, guid);
+        }
+        const res = await multi.exec();
```
Comment (Half-Shot) [inline src/Stores/RedisStorageProvider.ts:233]: There was a CPU hit from doing a lot of hits to redis at once. This at least avoids multiple transactions, and at least gives Redis the option of optimising many `lpos` requests. 
Annoyingly it doesn't have a way to ask for multiple list entries at once though.

File: src/feeds/parser.rs [hunk 4]
```diff
@@ -70,7 +70,8 @@ fn parse_channel_to_js_result(channel: &Channel) -> JsRssChannel {
                     .map(|f| f.value)
                     .or(item.link.clone())
                     .or(item.title.clone())
-                    .and_then(|f| hash_id(f).ok()),
+                    .and_then(|f| hash_id(f).ok())
+                    .map(|f| format!("md5:{}", f)),
```
Comment (Half-Shot) [inline src/feeds/parser.rs:74]: Cheap optimisation to do this Rust-side.

File: src/util/mod.rs [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        let since_the_epoch = start.duration_since(UNIX_EPOCH).unwrap().as_millis() as u64;
+
+        // We only need to check this once, as we won't be adding to the backoff queue
+        // as often as we pull from it.
+        if let Some(item) = self.backoff.first_entry() {
+            if *item.key() < since_the_epoch {
+                let v = item.remove();
+                self.queue.push_back(v);
+            }
+        }
```
Comment (AndrewFerr) [inline src/util/mod.rs:64]: Could you explain this part?  I'm reading it as "if the lowest backoff duration has passed, enqueue the URL that the backoff was for".  If that's the case, why is it safe to check that only once?  Is it because the queue is only for items whose backoff has expired & are thus always fair game to be fetched -- and because if no item has finished its backoff, this will return `None` and the caller will see no feeds available to poll?

File: src/util/mod.rs [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+        self.queue.pop_front()
+    }
+
+    #[napi]
+    pub fn remove(&mut self, item: String) -> bool {
+        // Remove from the queue
+        if let Ok(index) = self.queue.binary_search(&item) {
+            self.queue.remove(index);
+            return true;
```
Comment (AndrewFerr) [inline src/util/mod.rs:74]: Nitpick: be rusty
```suggestion
            true
```

File: src/util/mod.rs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            }
+        }
+
+        self.queue.pop_front()
+    }
+
+    #[napi]
+    pub fn remove(&mut self, item: String) -> bool {
+        // Remove from the queue
+        if let Ok(index) = self.queue.binary_search(&item) {
```
Comment (AndrewFerr) [inline src/util/mod.rs:72]: An item can be in either the queue or the backoff map, but not both, right?

File: src/util/mod.rs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+                    found_key = *key;
+                }
+            }
+            if found_key != 0 {
+                self.backoff.remove(&found_key);
+                return true;
+            }
+        }
+        // Always remove the duration on removal.
+        self.last_backoff_duration.remove(&item);
```
Comment (AndrewFerr) [inline src/util/mod.rs:91]: When would an item without a backoff have an entry in this?
Also, should this removal be done even when the entry does have a backoff?

File: src/util/mod.rs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        backoff_duration
+    }
+
+    #[napi]
+    pub fn length(&self) -> u32 {
+        self.queue.len() as u32
+    }
+
+    #[napi]
+    pub fn shuffle(&mut self) {
```
Comment (AndrewFerr) [inline src/util/mod.rs:136]: This could be made private

File: src/util/mod.rs (hunk 5 above)
Comment (Half-Shot) [reply src/util/mod.rs:64]: Sure, and please check my math here. We check the backoff queue once per `pull` call, which means we will pull one item (if available) and push it onto the "active" queue.

Each time we pull, we will do an action and then either push onto the active queue or the backoff queue. This means at worst case, we are only ever putting onto the backoff queue as often as we are pulling from it. In reality, it's unlikely that you are backing off every feed but if you were, you'd essentially just have a queue with a delay in it. The other thing to note here is that backed off items are inherently delayed, so that it's reasonably possible that we are `pull`ing more often than the items become "ready".

Now, it is possible that several items could all become ready within a few `ms` and the pull code isn't called exactly when they become ready. That's not really a problem, because we don't mind if items that are already failing are waiting a bit longer -- as long as they eventually get polled, they will enter the fast lane again. 

So to summarise why this is safe, we won't ever push to the backoff queue more often than we pull.

This is also efficient, because we're doing less work per pull (not iterating) and end users should notice no difference.

Does that make more sense?

File: src/util/mod.rs (hunk 7 above)
Comment (Half-Shot) [reply src/util/mod.rs:72]: yes

File: src/util/mod.rs (hunk 8 above)
Comment (Half-Shot) [reply src/util/mod.rs:91]: So if the item has never backed off, this is empty. If a entry that had previously failed had started working again, then we remove it. This map is purely to track that if we repeatedly back off an item, that we actually do increase the timeout rather than just timing out for the same duration each time.

File: src/util/mod.rs (hunk 6 above)
Comment (Half-Shot) [reply src/util/mod.rs:74]: apparently this was not allowed, maybe because it's not the last line in the function?

File: src/util/mod.rs (hunk 5 above)
Comment (AndrewFerr) [reply src/util/mod.rs:64]: Indeed it does, thanks!

**Review by AndrewFerr — APPROVED** [review]
